### PR TITLE
[SYCL-MLIR][NFC] Use `clang::Builtin` enum when possible

### PR DESCRIPTION
**Pull request by whitneywhtsang**
1. Used  `clang::Builtin` enum instead of string comparison when possible
2. Moved some builtin handling to `EmitBuiltinOps()`
3. Refactored `EmitBuiltinOps()`

Signed-off-by: Tsang, Whitney <whitney.tsang@intel.com>

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 1]
```diff
@@ -953,74 +953,246 @@ MLIRScanner::EmitBuiltinOps(clang::CallExpr *expr) {
...
-        return std::make_pair(
-            ValueCategory(builder.create<mlir::math::CosOp>(loc, args[0]),
-                          /*isReference*/ false),
-            true);
-      }
+  Optional<Value> V = None;
+  switch (expr->getBuiltinCallee()) {
+  case Builtin::BIceil: {
+    std::vector<mlir::Value> args;
+    for (auto a : expr->arguments()) {
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:961]: remove {}

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 2]
```diff
@@ -953,74 +953,246 @@ MLIRScanner::EmitBuiltinOps(clang::CallExpr *expr) {
...
+    }
+    V = builder.create<LLVM::CopySignOp>(loc, args[0], args[1]);
+  } break;
+  case Builtin::BI__builtin_exp2:
+  case Builtin::BI__builtin_exp2f:
+  case Builtin::BI__builtin_exp2l: {
+    std::vector<mlir::Value> args;
+    for (auto a : expr->arguments()) {
+      args.push_back(Visit(a).getValue(builder));
+    }
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:1027]: You could create a lambda to do this:
```
auto VisitArgs = [&]((clang::CallExpr &expr, std::vector<mlir::Value> &args){  
  for (auto a : expr.arguments()) {
      args.push_back(Visit(a).getValue(builder)); 
};

...

std::vector<mlir::Value> args;
switch() {
case Builtin::BI__builtin_exp2l:
   VisitArgs(*expr, args);
   V = builder.create<math::Exp2Op>(loc, args[0]);
   break;
}
```

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 3]
```diff
@@ -953,74 +953,246 @@ MLIRScanner::EmitBuiltinOps(clang::CallExpr *expr) {
                           /*isReference*/ false),
             true);
       }
-      if (sr->getDecl()->getIdentifier() && sr->getDecl()->getName() == "log") {
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:956]: Note where the code is moved pls.

Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:956]: Moved to line 978.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 4]
```diff
@@ -953,74 +953,177 @@ MLIRScanner::EmitBuiltinOps(clang::CallExpr *expr) {
...
-        std::vector<mlir::Value> args;
-        for (auto a : expr->arguments()) {
-          args.push_back(Visit(a).getValue(builder));
-        }
-        return std::make_pair(
-            ValueCategory(builder.create<mlir::math::LogOp>(loc, args[0]),
-                          /*isReference*/ false),
-            true);
-      }
-      if (sr->getDecl()->getIdentifier() &&
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:966]: Moved to line 965.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 5]
```diff
@@ -953,74 +953,177 @@ MLIRScanner::EmitBuiltinOps(clang::CallExpr *expr) {
...
-        std::vector<mlir::Value> args;
-        for (auto a : expr->arguments()) {
-          args.push_back(Visit(a).getValue(builder));
-        }
-        return std::make_pair(
-            ValueCategory(builder.create<mlir::math::SqrtOp>(loc, args[0]),
-                          /*isReference*/ false),
-            true);
-      }
-      if (sr->getDecl()->getIdentifier() &&
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:989]: Moved to line 973.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 6]
```diff
@@ -953,74 +953,177 @@ MLIRScanner::EmitBuiltinOps(clang::CallExpr *expr) {
...
-        std::vector<mlir::Value> args;
-        for (auto a : expr->arguments()) {
-          args.push_back(Visit(a).getValue(builder));
-        }
-        return std::make_pair(
-            ValueCategory(builder.create<mlir::math::ExpOp>(loc, args[0]),
-                          /*isReference*/ false),
-            true);
-      }
-      if (sr->getDecl()->getIdentifier() && sr->getDecl()->getName() == "sin") {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:1001]: Moved to line 982.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 7]
```diff
@@ -953,74 +953,177 @@ MLIRScanner::EmitBuiltinOps(clang::CallExpr *expr) {
...
-        for (auto a : expr->arguments()) {
-          args.push_back(Visit(a).getValue(builder));
-        }
-        return std::make_pair(
-            ValueCategory(builder.create<mlir::math::SinOp>(loc, args[0]),
-                          /*isReference*/ false),
-            true);
-      }
 
-      if (sr->getDecl()->getIdentifier() && sr->getDecl()->getName() == "cos") {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:1012]: Moved to line 969.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 8]
```diff
@@ -1146,45 +978,12 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
 
   if (auto *ic = dyn_cast<ImplicitCastExpr>(expr->getCallee()))
     if (auto *sr = dyn_cast<DeclRefExpr>(ic->getSubExpr())) {
-      if (sr->getDecl()->getIdentifier() &&
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:1149]: Moved to CFExpr.cc line 1094.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 9]
```diff
@@ -1146,45 +978,12 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
...
-        std::vector<mlir::Value> args = {
-            getLLVM(expr->getArg(0)), getLLVM(expr->getArg(1)),
-            getLLVM(expr->getArg(2)), /*isVolatile*/
-            builder.create<ConstantIntOp>(loc, false, 1)};
-
-        args[1] = builder.create<TruncIOp>(loc, builder.getI8Type(), args[1]);
-        builder.create<LLVM::MemsetOp>(loc, args[0], args[1], args[2], args[3]);
-        return ValueCategory(args[0], /*isReference*/ false);
-      }
-      if (sr->getDecl()->getIdentifier() &&
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:1172]: Moved to CFExpr.cc line 1114.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 10]
```diff
@@ -1146,45 +978,12 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
...
-           sr->getDecl()->getName() == "__builtin_memmove")) {
-        std::vector<mlir::Value> args = {
-            getLLVM(expr->getArg(0)), getLLVM(expr->getArg(1)),
-            getLLVM(expr->getArg(2)), /*isVolatile*/
-            builder.create<ConstantIntOp>(loc, false, 1)};
-        builder.create<LLVM::MemmoveOp>(loc, args[0], args[1], args[2],
-                                        args[3]);
-        return ValueCategory(args[0], /*isReference*/ false);
-      }
-      if (sr->getDecl()->getIdentifier() &&
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:1160]: Moved to CFExpr.cc line 1105.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 11]
```diff
@@ -1024,95 +926,25 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
...
+           BuiltinCallee == Builtin::BI__builtin_log10f ||
+           BuiltinCallee == Builtin::BI__builtin_log10l ||
            sr->getDecl()->getName() == "__nv_log10" ||
            sr->getDecl()->getName() == "__nv_log10f" ||
            sr->getDecl()->getName() == "__nv_log10l")) {
         mlir::Value V = getLLVM(expr->getArg(0));
         V = builder.create<math::Log10Op>(loc, V);
         return ValueCategory(V, /*isRef*/ false);
       }
-      if (sr->getDecl()->getIdentifier() &&
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:1037]: line 1037-1106 are moved to CFExpr.cc line 997.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 12]
```diff
@@ -837,134 +860,13 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
...
-        return ValueCategory(V, /*isRef*/ false);
-      }
       if (sr->getDecl()->getIdentifier() &&
           (sr->getDecl()->getName() == "__nv_fmodf")) {
         mlir::Value V = getLLVM(expr->getArg(0));
         mlir::Value V2 = getLLVM(expr->getArg(1));
         V = builder.create<mlir::LLVM::FRemOp>(loc, V.getType(), V, V2);
         return ValueCategory(V, /*isRef*/ false);
       }
-      if (sr->getDecl()->getIdentifier() &&
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:960]: Moved to CFExpr.cc line 991.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 13]
```diff
@@ -837,134 +860,13 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
...
-        mlir::Value BC = builder.create<BitcastOp>(loc, ITy, V);
-        auto ZeroV = builder.create<ConstantIntOp>(loc, 0, ITy);
-        V = builder.create<CmpIOp>(loc, CmpIPredicate::slt, BC, ZeroV);
-        auto postTy = Glob.getTypes()
-                          .getMLIRType(expr->getType())
-                          .cast<mlir::IntegerType>();
-        mlir::Value res = builder.create<ExtUIOp>(loc, postTy, V);
-        return ValueCategory(res, /*isRef*/ false);
-      }
-      if (sr->getDecl()->getIdentifier() &&
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:878]: line 879-952 are moved to CFExpr.cc line 1049.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 14]
```diff
@@ -837,134 +860,13 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
         mlir::Value res = builder.create<ExtUIOp>(loc, postTy, Eq);
         return ValueCategory(res, /*isRef*/ false);
       }
-      if (sr->getDecl()->getIdentifier() &&
-          (sr->getDecl()->getName() == "__builtin_isnormal")) {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:841]: Moved to line 660.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 15]
```diff
@@ -837,134 +860,13 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
...
-            builder.create<CmpFOp>(loc, CmpFPredicate::UGE, Abs, SmallestV);
-        V = builder.create<AndIOp>(loc, Eq, IsLessThanInf);
-        V = builder.create<AndIOp>(loc, V, IsNormal);
-        auto postTy = Glob.getTypes()
-                          .getMLIRType(expr->getType())
-                          .cast<mlir::IntegerType>();
-        mlir::Value res = builder.create<ExtUIOp>(loc, postTy, V);
-        return ValueCategory(res, /*isRef*/ false);
-      }
-      if (sr->getDecl()->getIdentifier() &&
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:864]: Moved to 681.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 16]
```diff
@@ -776,36 +827,8 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
             false);
       }
       if (sr->getDecl()->getIdentifier() &&
-          (sr->getDecl()->getName() == "__builtin_strlen" ||
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:779]: Moved to 634.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 17]
```diff
@@ -665,54 +755,16 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
       }
     }
 
-  if (auto *ic = dyn_cast<ImplicitCastExpr>(expr->getCallee()))
-    if (auto *sr = dyn_cast<DeclRefExpr>(ic->getSubExpr())) {
-      if (sr->getDecl()->getIdentifier() &&
-          sr->getDecl()->getName() == "__builtin_addressof") {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:671]: Moved to 693.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 18]
```diff
@@ -631,24 +631,114 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
   };
 
   switch (expr->getBuiltinCallee()) {
-  case clang::Builtin::BI__builtin_assume:
+  case Builtin::BI__builtin_strlen:
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:634]: Moved from line 779.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 19]
```diff
@@ -631,24 +631,114 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
...
+      module->push_back(Glob.getFunctions()[name]);
+    }
+
+    mlir::Value vals[] = {V0};
+    return ValueCategory(
+        builder.create<CallOp>(loc, Glob.getFunctions()[name], vals)
+            .getResult(0),
+        false);
+  }
+  case Builtin::BI__builtin_isnormal: {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:660]: Moved from line 841.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 20]
```diff
@@ -631,24 +631,114 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
...
+    mlir::Value IsNormal =
+        builder.create<CmpFOp>(loc, CmpFPredicate::UGE, Abs, SmallestV);
+    V = builder.create<AndIOp>(loc, Eq, IsLessThanInf);
+    V = builder.create<AndIOp>(loc, V, IsNormal);
+    auto postTy =
+        Glob.getTypes().getMLIRType(expr->getType()).cast<mlir::IntegerType>();
+    mlir::Value res = builder.create<ExtUIOp>(loc, postTy, V);
+    return ValueCategory(res, /*isRef*/ false);
+  }
+  case Builtin::BI__builtin_signbit: {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:681]: Moved from line 853.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 21]
```diff
@@ -631,24 +631,114 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
...
+    auto ITy = builder.getIntegerType(Ty.getWidth());
+    mlir::Value BC = builder.create<BitcastOp>(loc, ITy, V);
+    auto ZeroV = builder.create<ConstantIntOp>(loc, 0, ITy);
+    V = builder.create<CmpIOp>(loc, CmpIPredicate::slt, BC, ZeroV);
+    auto postTy =
+        Glob.getTypes().getMLIRType(expr->getType()).cast<mlir::IntegerType>();
+    mlir::Value res = builder.create<ExtUIOp>(loc, postTy, V);
+    return ValueCategory(res, /*isRef*/ false);
+  }
+  case Builtin::BI__builtin_addressof: {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGCall.cc:693]: Moved from line 671.

**Review by etiotto — APPROVED** [review]

**Pull request merged**